### PR TITLE
Convert ember-metal/transaction to typescript

### PR DESCRIPTION
**Pull request by csantero**
Also upgrades broccoli-typescript-compiler so as to pull in typescript 2.8, which I needed for the conditional types in `runInTransaction`.

### CODE REVIEW

**Review by rwjblue — COMMENTED** [review]
Will need to bump eslint-typescript-parser to ^15.0.0 to stay compatible with typescript@2.8.

File: packages/ember-metal/index.d.ts [hunk 1]
```diff
@@ -6,7 +6,17 @@ export const PROPERTY_DID_CHANGE: symbol;
...
-export function runInTransaction(context: any, methodName: string): any;
+export type MethodKey<T> = { [K in keyof T]: T[K] extends (() => void) ? K : never }[keyof T];
+export type RunInTransactionFunc = <T extends object, K extends MethodKey<T>>(
+  context: T,
+  methodName: K
+) => boolean;
+export type DidRenderFunc = (object: any, key: string, reference: any) => void;
+export type AssertNotRenderedFunc = (obj: object, keyName: string) => void;
+
+export const runInTransaction: RunInTransactionFunc;
```
Comment (rwjblue) [inline packages/ember-metal/index.d.ts]: I don't grok why we need this (along with the other lines above), I would have thought you could replace the preexisting `export function runInTransaction` line with `export { runInTransaction } from './lib/transaction';` and all of these types in this `index.d.ts` wouldn't need to be redefined. What am I missing?

Comment (csantero) [reply packages/ember-metal/index.d.ts]: @rwjblue You're right. I was over-thinking it. Fixed in latest commit.

**Comment by csantero** [comment]
> Will need to bump eslint-typescript-parser to ^15.0.0 to stay compatible with typescript@2.8.

Updated in latest commit

**Comment by csantero** [comment]
> Error: Navigation Timeout Exceeded: 30000ms exceeded

Travis seems to be timing out on one test suite.

**Comment by rwjblue** [comment]
restarted

**Pull request merged**